File: csharp/MapleStory.Test/KMS/History/TestGetCubeHistory.cs
using NUnit.Framework;
using MapleStory.OpenAPI.KMS;
using MapleStory.OpenAPI.Common;

namespace MapleStory.Test.KMS.History
{
    [TestFixture]
    public class TestGetCubeHistory
    {
        private static readonly string apiKey = EnvConfig.Get("API_KEY_KMS"); // Your API Key
        private static readonly MapleStoryAPI api = new MapleStoryAPI(apiKey);

        [Test, Description("success: GetCubeHistory")]
        public async Task GetCubeHistory()
        {
            int count = 1000;
            var response = await api.GetCubeHistory(count);
            Console.WriteLine(response.ToJson());
        }

        [Test, Description("success: GetCubeHistory with date")]
        public async Task GetCubeHistory_With_Date()
        {
            int count = 1000;
            var date = new DateTime(2025, 10, 23, 0, 0, 0);
            var response = await api.GetCubeHistory(count, date);
            Console.WriteLine(response.ToJson());
        }

        [Test, Description("success: GetCubeHistory [with cursor]")]
        public async Task GetCubeHistory_With_Cursor()
        {
            int count = 10;
            var date = new DateTime(2025, 10, 23, 0, 0, 0);
            var response = await api.GetCubeHistory(count, date);
            var nextCursor = response.NextCursor;
            Assert.That(nextCursor, Is.Not.Null);
            var nextResponse = await api.GetCubeHistory(count, nextCursor);
            Console.WriteLine(nextResponse.ToJson());
        }

        [Test, Description("fail: GetCubeHistory [throws OPENAPI00003 with invalid cursor]")]
        public void GetCubeHistory_With_Invalid_Cursor_Throws_OPENAPI00003()
        {
            int count = 1000;
            var invalidCursor = "invalid_cursor_123";
            var e = Assert.ThrowsAsync<MapleStoryAPIException>(async () => await api.GetCubeHistory(count, invalidCursor));
            Assert.That(e.ErrorCode, Is.EqualTo(MapleStoryAPIErrorCode.OPENAPI00003));
            Console.WriteLine($"{e.ErrorCode} {e.Message}");
        }

        [Test, Description("fail: GetCubeHistory [throws OPENAPI00004 with excessive count]")]
        public void GetCubeHistory_With_Excessive_Count_Throws_OPENAPI00004()
        {
            int count = 9999;
            var e = Assert.ThrowsAsync<MapleStoryAPIException>(async () => await api.GetCubeHistory(count));
            Assert.That(e.ErrorCode, Is.EqualTo(MapleStoryAPIErrorCode.OPENAPI00004));
            Console.WriteLine($"{e.ErrorCode} {e.Message}");
        }
    }
}

File: java/src/test/java/kms/history/TestGetCubeHistory.java
package kms.history;

import dev.spiralmoon.maplestory.api.common.MapleStoryApiErrorCode;
import dev.spiralmoon.maplestory.api.common.MapleStoryApiException;
import dev.spiralmoon.maplestory.api.kms.MapleStoryApi;
import dev.spiralmoon.maplestory.api.kms.dto.history.CubeHistoryResponseDTO;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.time.LocalDateTime;

import static org.assertj.core.api.Assertions.*;

@DisplayName("getCubeHistory")
public class TestGetCubeHistory {
    private static final String apiKey = System.getProperty("API_KEY_KMS"); // Your API Key
    private static final MapleStoryApi api = new MapleStoryApi(apiKey);

    @Test
    @DisplayName("success: getCubeHistory")
    void getCubeHistory() {
        int count = 1000;
        CubeHistoryResponseDTO response = api.getCubeHistory(count).join();
        System.out.println(response.toString());
    }

    @Test
    @DisplayName("success: getCubeHistory with date")
    void getCubeHistory_with_date() {
        int count = 1000;
        LocalDateTime date = LocalDateTime.of(2025, 10, 23, 0, 0);
        CubeHistoryResponseDTO response = api.getCubeHistory(count, date).join();
        System.out.println(response.toString());
    }

    @Test
    @DisplayName("success: getCubeHistory with cursor")
    void getCubeHistory_with_cursor() {
        int count = 10;
        LocalDateTime date = LocalDateTime.of(2025, 10, 23, 0, 0);
        CubeHistoryResponseDTO response = api.getCubeHistory(count, date).join();
        String nextCursor = response.getNextCursor();
        assertThat(nextCursor).isNotNull();

        CubeHistoryResponseDTO nextResponse = api.getCubeHistory(count, nextCursor).join();
        System.out.println(nextResponse.toString());
    }

    @Test
    @DisplayName("fail: getCubeHistory with invalid cursor throw OPENAPI00003")
    void getCubeHistory_with_invalid_cursor() {
        int count = 1000;
        String invalidCursor = "invalid_cursor_123";
        assertThatThrownBy(() -> api.getCubeHistory(count, invalidCursor).join())
                .hasCauseInstanceOf(MapleStoryApiException.class)
                .satisfies(e -> {
                    MapleStoryApiException apiException = (MapleStoryApiException) e.getCause();
                    assertThat(apiException.getErrorCode()).isEqualTo(MapleStoryApiErrorCode.OPENAPI00003);
                    System.out.println(apiException.getErrorCode() + " " + apiException.getMessage());
                });
    }

    @Test
    @DisplayName("fail: getCubeHistory with excessive count throw OPENAPI00004")
    void getCubeHistory_with_excessive_count() {
        int count = 9999;
        assertThatThrownBy(() -> api.getCubeHistory(count).join())
                .hasCauseInstanceOf(MapleStoryApiException.class)
                .satisfies(e -> {
                    MapleStoryApiException apiException = (MapleStoryApiException) e.getCause();
                    assertThat(apiException.getErrorCode()).isEqualTo(MapleStoryApiErrorCode.OPENAPI00004);
                    System.out.println(apiException.getErrorCode() + " " + apiException.getMessage());
                });
    }
}
